dimod/compatibility23.py

```diff
@@ -35,3 +43,32 @@ def getargspec(f):
         # FullArgSpec(args, varargs, varkw, defaults, kwonlyargs, kwonlydefaults, annotations)
 
         return _ArgSpec(argspec.args, argspec.varargs, argspec.varkw, argspec.defaults)
+
+    # Based on an answer https://stackoverflow.com/a/34757114/8766655
+    # by kindall https://stackoverflow.com/users/416467/kindall
+    class SortKey(object):
+        def __init__(self, obj):
+            self.obj = obj
+            self.name = type(self.obj).__name__
+
+        def __lt__(self, other):
+            try:
+                return self.obj < other.obj
+            except TypeError:
+                pass
+
+            if self.name == other.name:
+
+                if isinstance(self.obj, abc.Sequence):
+                    # this case happens when there are two sequences of the same type
+                    # that have nested objects that python3 cannot compare
+                    return [SortKey(v) for v in self.obj] < [SortKey(v) for v in other.obj]
+
+                # if they are of the same type but they failed the original
+                # try-catch block and they are not a sequence then we can't
+                # resolve the order (this happens for instance with dicts which
+                # python2 can sort but not in python3)
+                msg = "cannot sort types {!r} and {!r}"
+                raise TypeError(msg.format(self.obj.__name__, other.obj.__name__))
```

Comment:
Why not use `self.name` and `other.name` here? Otherwise you get stuff like `AttributeError: 'dict' object has no attribute '__name__'` before you can raise your error.

Reply:
oops, that is what I meant. Will fix